torch/csrc/jit/tensorexpr/graph_opt.cpp

```diff
@@ -0,0 +1,164 @@
+#include <torch/csrc/jit/tensorexpr/graph_opt.h>
+
+#include <torch/csrc/jit/passes/tensorexpr_fuser.h>
+#include <torch/csrc/jit/tensorexpr/kernel.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+// Move the given user of `aten::cat` op to its inputs.
+Node* moveCatAfterUse(Node* cat, Node* user, std::shared_ptr<Graph> subgraph) {
+  // Example IR:
+  //   %1 = ...
+  //   %2 = ...
+  //   %3 = prim::ListConstruct(%1, %2)
+  //   %4 = aten::cat(%3, ...)
+  //   %5 = aten::relu(%4)
+  //   return (%5)
+  //
+  // To be transformed to:
+  //   %1 = ...
+  //   %2 = ...
+  //   %5.1 = aten::relu(%1)
+  //   %5.2 = aten::relu(%2)
+  //   %3 = prim::ListConstruct(%5.1, %5.2)
+  //   %4 = aten::cat(%3, ...)
+  //   return (%4)
+
+  TORCH_INTERNAL_ASSERT(cat->output()->hasUses());
+  TORCH_INTERNAL_ASSERT(cat->output()->uses().size() == 1);
+  TORCH_INTERNAL_ASSERT(cat->input(0)->node()->kind() == prim::ListConstruct);
+  auto cat_list = cat->input(0)->node();
+  auto cat_inputs = cat_list->inputs();
+
+  auto user_tensor_type = user->output()->type()->cast<c10::TensorType>();
+  TORCH_INTERNAL_ASSERT(user_tensor_type);
+  std::unordered_map<Value*, Value*> new_cat_inputs;
+  for (auto inp : cat_inputs) {
+    auto new_cat_input = subgraph->createClone(
+        user, [&](Value* k) { return (k == cat->output()) ? inp : k; });
+    // Since we are cloning user, its result should be the same scalar type
+    // as the user. But the dims should correspond to that of the input.
+    auto input_tensor_type = inp->type()->cast<c10::TensorType>();
+    TORCH_INTERNAL_ASSERT(input_tensor_type);
+    auto new_input_type =
+        input_tensor_type->withScalarType(user_tensor_type->scalarType());
+    new_cat_input->output()->setType(new_input_type);
+    new_cat_input->insertBefore(cat_list);
+    new_cat_inputs[inp] = new_cat_input->output();
+  }
+  auto new_cat_list = subgraph->createClone(
+      cat_list, [&](Value* k) { return new_cat_inputs[k]; });
+  new_cat_list->insertBefore(cat);
+  auto new_cat = subgraph->createClone(cat, [&](Value* k) {
+    return (k == cat_list->output()) ? new_cat_list->output() : k;
+  });
+  new_cat->output()->setType(user_tensor_type);
+  new_cat->insertBefore(cat);
+
+  user->output()->replaceAllUsesWith(new_cat->output());
+  user->destroy();
+
+  TORCH_INTERNAL_ASSERT(!cat->output()->hasUses());
+  cat->destroy();
+
+  if (!cat_list->output()->hasUses()) {
+    cat_list->destroy();
+  }
```

Comment:
Maybe a question for @eellison  too: do we actually have to explicitly destroy the cat/listconstruct, or will those get cleaned up by dce or something?